### DISCUSSION TITLE
Use swc’s ast for analytics

### DISCUSSION THREAD

**Discussion by filipw01** [swc-project/swc#5396] [Q&A]
Hey, I want to create some usage statistics for my project (ts/tsx/js/jsx). I know I could write a plugin, but I don’t need any additional transforms for my code, only an ast and my visitor to run on it. Should I use some swc ast crate directly or is there a way to run swc without any transforms, but with my plugin? Or is there another preferable way to do this? I don’t have any experience with swc so this might sound dumb, but I would value any response nonetheless

**Comment by Austaras** [comment]
You could write your own rust program and only use swc's parser and resolver

**Reply by filipw01** [reply]
Thank you, I'll try that

**Reply by filipw01** [reply]
I'm a bit lost. I got it parsed, but I don't know how to implement a visitor/resolver I saw some older posts, but it seems some API has changed because `Module` no longer has `visit` method on it
```rust
extern crate swc_common;
extern crate swc_ecma_parser;
extern crate swc_ecma_ast;
extern crate swc_ecma_visit;

use std::path::Path;
use swc_common::sync::Lrc;
use swc_common::{
    errors::{ColorConfig, Handler},
    SourceMap,
};
use swc_ecma_parser::{lexer::Lexer, Parser, StringInput, Syntax, TsConfig};

fn main() {
    let cm: Lrc<SourceMap> = Default::default();
    let handler =
        Handler::with_tty_emitter(ColorConfig::Auto, true, false, Some(cm.clone()));

    let fm = cm
        .load_file(Path::new("src/file.tsx"))
        .expect("failed to load file.tsx");
    let lexer = Lexer::new(
        Syntax::Typescript(TsConfig {
            tsx: true,
            ..Default::default()
        }),
        Default::default(),
        StringInput::from(&*fm),
        None,
    );
    let mut parser = Parser::new_from(lexer);

    for e in parser.take_errors() {
        e.into_diagnostic(&handler).emit();
    }

    let m = parser
        .parse_module()
        .map_err(|e| {
            // Unrecoverable fatal error occurred
            e.into_diagnostic(&handler).emit()
        })
        .expect("failed to parser module");

    println!("{:#?}", m);
}
```
